prov/efa/src/rdm/efa_rdm_ope.c

```diff
@@ -449,68 +450,105 @@ size_t efa_rdm_txe_max_req_data_capacity(struct efa_rdm_ep *ep, struct efa_rdm_o
+			*pkt_entry_cnt = available_tx_pkts;
+		assert(*pkt_entry_cnt > 0 && *pkt_entry_cnt <= EFA_RDM_EP_MAX_WR_PER_IBV_POST_SEND);
+		for (i = 0; i < *pkt_entry_cnt - 1; ++i)
+			pkt_entry_data_size_vec[i] = ope->ep->max_data_payload_size;
+
+		remainder = ope->window - (*pkt_entry_cnt - 1) * ep->max_data_payload_size;
+		assert(remainder > 0);
+		pkt_entry_data_size_vec[i] = MIN(remainder, ep->max_data_payload_size);
+		return 0;
 	}
 
-	max_req_data_capacity_aligned = (max_req_data_capacity & ~(memory_alignment - 1));
-	num_req = (mulreq_total_data_size - 1)/max_req_data_capacity_aligned + 1;
-	txe->max_req_data_size = ofi_get_aligned_size((mulreq_total_data_size - 1)/num_req + 1, memory_alignment);
-	if (txe->max_req_data_size > max_req_data_capacity_aligned)
-		txe->max_req_data_size = (max_req_data_capacity_aligned);
-}
+	if (rxr_pkt_type_is_medium(pkt_type) || rxr_pkt_type_is_runt(pkt_type)) {
+		if(rxr_pkt_type_is_runt(pkt_type))
+			efa_rdm_txe_set_runt_size(ep, ope);
 
-/**
- * @brief return number of REQ packets needed to send a message using mulit-req protocol
- *
- * @param[in]		txe		txe with information of the message
- * @param[in]		pkt_type		packet type of the mulit-req protocol
- * @return			number of REQ packets
- */
-size_t efa_rdm_txe_num_req(struct efa_rdm_ope *txe, int pkt_type)
-{
-	assert(rxr_pkt_type_is_mulreq(pkt_type));
-	assert(txe->max_req_data_size);
+		total_pkt_entry_data_size = efa_rdm_ope_mulreq_total_data_size(ope, pkt_type);
+		assert(total_pkt_entry_data_size);
+
+		single_pkt_entry_max_data_size = efa_rdm_txe_max_req_data_capacity(ep, ope, pkt_type);
+		assert(single_pkt_entry_max_data_size);
+
+		if (efa_mr_is_cuda(ope->desc[0])) {
+			if (ep->sendrecv_in_order_aligned_128_bytes)
+				memory_alignment = EFA_RDM_IN_ORDER_ALIGNMENT;
+			else
+				memory_alignment = EFA_RDM_CUDA_MEMORY_ALIGNMENT;
+		}
+
+		*pkt_entry_cnt = (total_pkt_entry_data_size - 1) / single_pkt_entry_max_data_size + 1;
+		if (*pkt_entry_cnt > available_tx_pkts)
+			return -FI_EAGAIN;
 
-	size_t total_size = efa_rdm_ope_mulreq_total_data_size(txe, pkt_type);
+		/* when sending multiple packets, it is more performant that the data size of each packet are close
+		 * to achieve that, we calculate the single packet size
+		 */
+		single_pkt_entry_data_size = (total_pkt_entry_data_size - 1) / *pkt_entry_cnt + 1;
 
-	return (total_size - txe->bytes_sent - 1)/txe->max_req_data_size + 1;
+		/* each packet must be aligned */
+		single_pkt_entry_data_size = single_pkt_entry_data_size & ~(memory_alignment - 1);
+
+		*pkt_entry_cnt = (total_pkt_entry_data_size - 1) / single_pkt_entry_data_size + 1;
+
+		for (i = 0; i < *pkt_entry_cnt - 1; ++i)
+			pkt_entry_data_size_vec[i] = single_pkt_entry_data_size;
+
+		remainder = total_pkt_entry_data_size - (*pkt_entry_cnt - 1) * single_pkt_entry_data_size;
+		pkt_entry_data_size_vec[*pkt_entry_cnt - 1] = remainder;
+		return 0;
+	}
+
+	*pkt_entry_cnt = 1;
+	pkt_entry_data_size_vec[0] = -1;
```

Comment:
does this mean an error or this vector isn't needed? Need a comment

Reply:
added a comment in new revision.